pyfr/integrators/dual/phys/steppers.py

```diff
@@ -4,22 +4,127 @@
+    def _finalize_step(self):
+        pass
+
+
+class BaseBDFStepper(BaseDualStepper):
+    nstages = 1
+    stage_nregs = 0
+    c = [1]
+
+    @property
+    def stepper_nregs(self):
+        return len(self.stepper_static_coeffs) - 1
+
+    def _finalize_step(self):
+        psnregs = self.pseudointegrator.pseudo_stepper_nregs
+        snregs = self.pseudointegrator.stepper_nregs
+
+        # Rotate the source registers to the right by one
+        self.pseudointegrator._regidx[psnregs:psnregs + snregs] = (
+            self.pseudointegrator._stepper_regidx[-1:]
+            + self.pseudointegrator._stepper_regidx[:-1]
+        )
+
+        # Copy the current soln into the first source register
+        self.pseudointegrator._add(
+            0, self.pseudointegrator._stepper_regidx[0],
+            1, self.pseudointegrator._idxcurr
+        )
+
+    @property
+    def stepper_coeffs(self):
+        return [1] + [sc/self._dt for sc in self.stepper_static_coeffs]
+
+
+class DualBDF2Stepper(BaseBDFStepper):
     stepper_name = 'bdf2'
     stepper_order = 2
-    stepper_coeffs = [-1.5, 2.0, -0.5]
+    stepper_static_coeffs = [-1.5, 2.0, -0.5]
 
 
-class DualBDF3Stepper(BaseDualStepper):
+class DualBDF3Stepper(BaseBDFStepper):
     stepper_name = 'bdf3'
     stepper_order = 3
-    stepper_coeffs = [-11.0/6.0, 3.0, -1.5, 1.0/3.0]
+    stepper_static_coeffs = [-11.0/6.0, 3.0, -1.5, 1.0/3.0]
 
 
-class DualBackwardEulerStepper(BaseDualStepper):
+class DualBackwardEulerStepper(BaseBDFStepper):
     stepper_name = 'backward-euler'
     stepper_order = 1
-    stepper_coeffs = [-1.0, 1.0]
+    stepper_static_coeffs = [-1.0, 1.0]
+
+
+class BaseDIRKStepper(BaseDualStepper):
+    stepper_nregs = 1
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        if self.fsal:
+            self.b = self.a[-1]
+
+        self.c = [sum(row) for row in self.a]
+
+    @property
+    def stage_nregs(self):
+        return self.nstages
+
+    @property
+    def stepper_coeffs(self):
+        csn = self.currstg
+        return [self.a[csn][csn], -1/self._dt, 1/self._dt] + self.a[csn][:csn]
+
+    def _finalize_step(self):
+        if not self.fsal:
+            self._obtain_solution()
+
+        # Copy the new solution into stepper register
+        self.pseudointegrator._add(
+            0, self.pseudointegrator._stepper_regidx[0],
+            1, self.pseudointegrator._idxcurr
+        )
+
+    def _obtain_solution(self):
+        consts = [0, 1] + [bt*self._dt for bt in self.b]
+        regidxs = ([self.pseudointegrator._idxcurr
+                    + self.pseudointegrator._stepper_regidx[0]]
+                   + self.pseudointegrator._stage_regidx)
+        self.pseudointegrator._add(consts, regidxs,
+                                   subdims=self.pseudointegrator._subdims)
+
+
+class SDIRK33Stepper(BaseDIRKStepper):
+    stepper_name = 'sdirk33'
+    nstages = 3
+    fsal = True
+
+    _a_lam = 0.4358665215084590
```

Comment:
Is there a closed form for this?

Reply:
There is for the other constant but this is given as it is in Carpenter and Kennedy's book

Reply:
Does not appear to have an obvious form: http://wayback.cecm.sfu.ca/cgi-bin/isc/lookup?number=0.4358665215084590&lookup_type=simple